runbook/api/v1/writer/runbook_.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2016: Mirantis Inc.
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import elasticsearch
+import flask
+import jsonschema
+
+from runbook.api import utils
+from runbook import storage
+
+API_TYPE = "writer"
+bp = flask.Blueprint("runbooks", __name__)
+
+
+def get_blueprints():
+    return [["/region", bp]]
+
+
+RUNBOOK_SCHEMA = {
+    "type": "object",
+    "$schema": "http://json-schema.org/draft-04/schema",
+    "properties": {
+        "name": {"type": "string"},
+        "description": {"type": "string"},
+        "type": {"type": "string"},
+        "runbook": {"type": "string"},
+    },
+    "required": ["name", "description", "runbook", "type"],
+    "additionalProperties": False
+}
+
+
+def _convert(hit):
+    body = {k: v for k, v in hit["_source"].items()}
+    body["_id"] = hit["_id"]
+    return body
+
+
+@bp.route("/<region>/runbooks", methods=["POST"])
+@utils.check_regions(API_TYPE)
+def handle_runbooks(region):
+    es = storage.get_elasticsearch(API_TYPE)
+    index_name = "ms_runbooks_{}".format(region)
+
+    runbook = flask.request.get_json(silent=True)
+    try:
+        jsonschema.validate(runbook, RUNBOOK_SCHEMA)
+    except jsonschema.ValidationError as e:
+        # NOTE(kzaitsev): jsonschema exception has really good unicode
+        # error representation
+        return flask.jsonify(
+            {"error": u"{}".format(e)}), 400
```

Comment:
it is generic, but writing it as a decorator — I believe would only hurt readability. it's only used in PUT branch of the 2d endpoint, so this decorator would need to somehow handle that. Looks like more code to me